### DISCUSSION TITLE
In process certificate from KeyVault

### DISCUSSION THREAD

**Discussion by wonderplayer** [pnp/pnpcore#916] [Q&A] [answered]
Hello,

I've seen a very cool approach [here](https://github.com/pnp/pnpcore/blob/dev/samples/Demo.AzureFunction.OutOfProcess.AppOnly/Program.cs) to either use local thumbprint of a certificate or get a KeyVault reference.

I successfully reproduced it in .NET 6 isolated and was happy. But then I wanted to add same behaviour to .NET 6 in-process and I just can't get it to work.

Do you have any suggestions how to do it in .NET 6 in-process - with local and key vault flows?

Thank you!

**Comment by jansenbe** [comment]
@wonderplayer : I would assume this also works in .NET 6 in process: what error do you get?

**Reply by wonderplayer** [reply]
Yeah, me too. I'm pretty confident that I've set them up equivalently 😄 
I'm getting:
```
[2022-07-14T12:49:21.178Z] Executed 'HttpTrigger1' (Failed, Id=b07a92fd-3a53-4f7b-9404-d6f88f8e6d9a, Duration=12053ms)
[2022-07-14T12:49:21.180Z] Anonymously Hosted DynamicMethods Assembly: Object reference not set to an instance of an object.
[2022-07-14T12:49:24.554Z] An unhandled host error has occurred.
[2022-07-14T12:49:24.556Z] Anonymously Hosted DynamicMethods Assembly: Object reference not set to an instance of an object.
```


Repositories:

- Isolated: https://github.com/wonderplayer/afa6-isolated
- In process: https://github.com/wonderplayer/afa6-inprocess

**Reply by wonderplayer** [reply]
@jansenbe did you have a chance to look at this?

**Reply by jansenbe** [reply]
@wonderplayer : I've not, been OOF for a long time and know catching up on things.

**Reply by lroos** [reply, marked as answer]
I had another approach that's worked for loading certs from key vault - use the DefaultCertificateLoader from Microsoft.Identity.Web. Add a client cert config section similar to described here https://github.com/AzureAD/microsoft-identity-web/wiki/Using-certificates and bind it to a CertificateDescription object, `new DefaultCertificateLoader().LoadIfNeeded(clientCert)` will determine if it's a Key Vault source and it uses `DefaultAzureCredential` to authenticate to Key Vault. In an Azure environment that should use the `ManagedIdentityCredential`, which should be the managed identity of your function app. It's a really nice abstraction.

**Reply by wonderplayer** [reply]
@lroos Thank you for suggestion!
Does it mean that, if I put in Azure App Settings certificate configuration and in my local environment local file configuration with password, then `DefaultCertificateLoader` will have loaded certificate, and only thing I will need to do is to get correct X509Certificate from it?

**Reply by lroos** [reply]
Pretty much yes, like:
```cs
_certificateLoader.LoadIfNeeded(_options.ClientCert); // ClientCert is of type CertificateDescription
var cert = _options.ClientCert.Certificate ?? throw ...
```
For local dev you could use StoreType "Path" for .pfx or "StoreWithDistinguishedName" on for your local cert store. But it will also work if you point it to Key Vault during dev; the [DefaultAzureCredential](https://learn.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet) will pick up your Azure Service Auth credential if you're in Visual Studio or VS Code, or your `az login` credential if run from command line.
